### PR TITLE
feat(flags): Add distinct id and group key matching

### PR DESCRIPTION
**Pull request by neilkakkar**
## Problem

Split out from https://github.com/PostHog/posthog/pull/19792/ -> add group and distinct id targeting to flags

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

👀

### CODE REVIEW

**Review by mariusandra — COMMENTED** [review]
Is it possible to merge `metadataTaxonomicGroupToPropertyFilterType` and `optionsFromProp`? So that you'd pass along something like:

```
const taxonomicOptions: TaxonomicFilterProps['optionsFromProp'] = {
    [TaxonomicFilterGroupType.Metadata]: [{ name: 'distinct_id', return_type: PropertyFilterType.Person }],
}
```

This would make the metadata field much more _meta_, and avoid passing through one large and very specific key.

Or is that [complicated](https://m.youtube.com/watch?v=RXJKdh1KZ0w&) somehow? 🤔

**Comment by neilkakkar** [comment]
Slightly complicated, leads to a bigger refactor I'm not _yet_ keen on, because propertyType is determined per taxonomicGroup, not specific selected items inside the group 🫠 -> https://github.com/PostHog/posthog/blob/metadata-flags/frontend/src/lib/components/PropertyFilters/components/taxonomicPropertyFilterLogic.ts#L92

**Comment by neilkakkar** [comment]
... actually, doesn't seem too bad? Let me try it 👀 , would make things nicer for sure, and prevent the 'metadata needs to be single type' in future problem

**Comment by neilkakkar** [comment]
Ok, done, this is nice 🙌

File: frontend/src/lib/components/PropertyFilters/components/taxonomicPropertyFilterLogic.ts [hunk 1]
```diff
@@ -129,9 +133,12 @@ export const taxonomicPropertyFilterLogic = kea<taxonomicPropertyFilterLogicType
...
                         PropertyOperator.Exact
 
+                    const isGroupNameFilter = taxonomicGroup.type.startsWith(TaxonomicFilterGroupType.GroupNamesPrefix)
+                    // :TRICKY: When we have a GroupNamesPrefix taxonomic filter, selecting the group name
+                    // is the equivalent of selecting a property value
                     const property: AnyPropertyFilter = {
-                        key: propertyKey.toString(),
-                        value: null,
+                        key: isGroupNameFilter ? '$group_key' : propertyKey.toString(),
+                        value: isGroupNameFilter ? propertyKey.toString() : null,
```
Comment (mariusandra) [inline frontend/src/lib/components/PropertyFilters/components/taxonomicPropertyFilterLogic.ts:141]: This override is the last thing that raises eyebrows... Happy to let is pass though, the impact is very minimal.

The bigger UX sadness here, which can also be for later, is that once you select an organization, the filter says "Group key", not "Organization key".



The title there can probably be field, but up to you if you want to...

**Review by mariusandra — APPROVED** [review]
Nice.

File: frontend/src/lib/components/PropertyFilters/components/taxonomicPropertyFilterLogic.ts (hunk 1 above)
Comment (neilkakkar) [reply frontend/src/lib/components/PropertyFilters/components/taxonomicPropertyFilterLogic.ts:141]: Want to keep it is `group_key` for now, given it's in context of matching org(s), and this is the actual field when, say, calling `$groupIdentify()`. Will revisit though if this raises confusion

**Pull request merged**